File: pagr_drm.py
import json
import urllib.request
import requests
import secrets
import sys
import hashlib
import uuid

DEBUG = True
URL = "http://localhost:3000/pagr_drm.php"
INVALID_RESPONSES = [400, 403, 404]

def PrepareJson(pairs):
    data = {}
    for key, value in pairs:
        data[key] = value
    return json.dumps(data)

def Post(data):
    Debug(f"POST data: {data}")

    headers = {'Content-Type': 'application/json'}
    url = URL

    try:
        response = requests.post(url, data=data, headers=headers, timeout=10)
        response.raise_for_status()

        responseText = response.text
        responseStatusCode = response.status_code

    except requests.exceptions.ConnectionError:
        Debug(f"ConnectionError: \"{URL}\" is either down, or unreachable to the network.")
        responseText = "{failure}"
        responseStatusCode = "0"

    Debug(f"{responseStatusCode}: {responseText}")
    return responseText, responseStatusCode

def DRMKeyGenerator():
    return secrets.token_hex(32)

def HardwareID():
    mac = uuid.getnode()
    if (mac >> 40) % 2:
        raise ValueError("MAC address is not available or cannot be used.")
    mac_str = ':'.join(['{:02x}'.format((mac >> ele) & 0xff) for ele in range(0,8*6,8)][::-1])
    hardware_id = hashlib.sha256(mac_str.encode()).hexdigest()
    
    return hardware_id

def VerifyResponse(responseStatusCode):
    if responseStatusCode in INVALID_RESPONSES:
        Name()
        print("[Failure] Either the DRM server is down, or the client Internet gateway is unreachable.")
        sys.exit(1)

def Check(drm_key):
    hardwareID = HardwareID()
    json = PrepareJson([["drm_key", drm_key],["hardware_id", hardwareID],])
    # Once this is posted with the DRM key and hardware ID, if it's not
    # registered, register it on the PHP side.
    # If registration is successful, a
    # {10} will be outputted. If an {11} is outputted, there was an error with registration
    responseText, responseStatusCode = Post(json)

    VerifyResponse(responseStatusCode)

    # NOTE: Might have to convert the responseText to JSON or something.

    match responseText:
        #--------------#
        # Registration #
        #--------------#
        case "{1}":
            Debug(f"{responseText}: Correct key \"{drm_key}\", already registered \"{hardwareID}\". Approved.")
            return
        case "{10}":
            Debug(f"{responseText}: Correct key \"{drm_key}\", registration complete \"{hardwareID}\". Approved.")
            return
        case "{11}":
            Debug(f"{responseText}: Correct key \"{drm_key}\", registration failed \"{hardwareID}\". Denied.")
            sys.exit(1)

        #-----------------------------#
        # Overregistered (monkaOmega) #
        #-----------------------------#
        case "{2}":
            Debug(f"{responseText}: Correct key \"{drm_key}\", too many device registrations \"{hardwareID}\". Denied!")
            Debug(f"Possible piracy detected. This incident will be reported.")
            sys.exit(1)

        case "{0}":
            Debug(f"{responseText}: Failed to reach server!")
            sys.exit(1)

        case _:
            print(f"Unknown case detected (multiline):\n\t{responseText}")
            sys.exit(1)

def Debug(message):
    if DEBUG:
        print(f"[DEBUG] {message}")

def Name():
    print("pagr_drm: Protect your software.")

File: pagr_drm.php
<?php
// pagr_drm.php

header('Content-Type: application/json');

// Database credentials
$servername = "localhost";
$username = "eO3EPtUBC0j896YExiVp"; // Replace with your DB username
$password = "uON2dlBF9sHWrnap4OjH@"; // Replace with your DB password
$dbname = "pagr_drm";

// Get the POST data
$input = file_get_contents('php://input');
$data = json_decode($input, true);

if (!isset($data['drm_key']) || !isset($data['hardware_id'])) {
    http_response_code(400);
    echo json_encode(["error" => "Invalid request"]);
    exit();
}

$drm_key = $data['drm_key'];
$hardware_id = $data['hardware_id'];

// Create connection
$conn = new mysqli($servername, $username, $password, $dbname);

// Check connection
if ($conn->connect_error) {
    http_response_code(500);
    echo json_encode(["error" => "Database connection failed"]);
    exit();
}

// Escape input data to prevent SQL injection
$drm_key_esc = $conn->real_escape_string($drm_key);
$hardware_id_esc = $conn->real_escape_string($hardware_id);

// Validate drm_key
$sql = "SELECT max_devices FROM drm_keys WHERE drm_key = '$drm_key_esc'";
$result = $conn->query($sql);

if ($result->num_rows == 0) {
    // Invalid DRM key
    http_response_code(403);
    echo json_encode(["error" => "Invalid DRM key"]);
    exit();
}

$row = $result->fetch_assoc();
$max_devices = $row['max_devices'];

// Check if hardware_id is already registered
$sql = "SELECT * FROM registrations WHERE drm_key = '$drm_key_esc' AND hardware_id = '$hardware_id_esc'";
$result = $conn->query($sql);

if ($result->num_rows > 0) {
    // Already registered
    echo json_encode(["message" => "Already registered", "status" => "approved"]);
    exit();
}

// Check if the maximum number of devices has been reached
$sql = "SELECT COUNT(*) as device_count FROM registrations WHERE drm_key = '$drm_key_esc'";
$result = $conn->query($sql);
$row = $result->fetch_assoc();
$device_count = $row['device_count'];

if ($device_count >= $max_devices) {
    // Too many devices registered
    echo json_encode(["error" => "Too many devices registered", "status" => "denied"]);
    exit();
}

// Register the new hardware_id
$sql = "INSERT INTO registrations (drm_key, hardware_id) VALUES ('$drm_key_esc', '$hardware_id_esc')";

if ($conn->query($sql) === TRUE) {
    // Registration successful
    echo json_encode(["message" => "Registration complete", "status" => "approved"]);
    exit();
} else {
    // Registration failed
    echo json_encode(["error" => "Registration failed", "status" => "denied"]);
    exit();
}

$conn->close();
?>
